Rewrite wsbroadcast to not rely on access to shared redis (wsrelay)

shanemcd opened the pull request:
Refs #13319

The system currently looks something like this:

![wsbroadcast-2](https://user-images.githubusercontent.com/773186/200655678-5fe903e1-3b56-4e6d-b466-6082d56a33ce.png)

This design is not optimal for a couple reasons:
- Currently wsbroadcast dials out to every other node and consumes their broadcast messages. The current [BroadcastConsumer](https://github.com/ansible/awx/blob/15964dc3959472950db23ed6463c7f4e1978192c/awx/main/consumers.py#L93) emits messages that are sent to a Django Channels group called called broadcast-group_send. This means that [emit_channel_notification](https://github.com/ansible/awx/blob/15964dc3959472950db23ed6463c7f4e1978192c/awx/main/consumers.py#L219-L226) needs to write to the same instance of Redis that Daphne is reading from.
- We broadcast every message that the system emits. Regardless of whether or not there is a client.

I spent some time thinking through how we might be able to solve both problems at the same time and came up with this:

![wsrelay](https://user-images.githubusercontent.com/773186/200655868-80414981-a279-467b-ac64-07f540c96172.png)

shanemcd committed:
Checkpoint

AlanCoding commented on tools/docker-compose/supervisor.conf:
```diff
@@ -8,31 +8,25 @@ command = make dispatcher
-stderr_logfile=/dev/stderr
-stderr_logfile_maxbytes=0
+stdout_events_enabled = true
+stderr_events_enabled = true
```
Testing this change locally, this seems to keep supervisord logs like

```
2022-11-09 09:34:47,046 INFO spawned: 'fail-slowly' with pid 1258318
```

but stops sending the subprocess standard out to the supervisord standard out. For docker-compose I don't think that's what we want? Anyway, this is fairly unrelated to the feature intent here.

shanemcd replied:
Hmm I am not sure I'm following. The intent here is to send all logs to stdout in our dev env rather than /var/log/supervisor/.

AlanCoding replied:
There are logs from supervisor (like what I pasted) and standard out from the processes that supervisord is managing. Yes, we're on the same page that both types of logs should go to stdout in the dev environment. My comment was saying that standard out from managed processes (like dispatcher/uwsgi/etc.) do not with those settings. When it says "stdout_events_enabled", I think it means supervisor events. The docs mention something about capture mode playing a role, but I don't understand it.

AlanCoding commented on awx/conf/settings.py:
```diff
@@ -187,6 +184,18 @@ def get_settings_to_cache(registry):
     return dict([(key, SETTING_CACHE_NOTSET) for key in get_writeable_settings(registry)])
 
 
+# HACK: runs in thread in order to work in an asyncio context
```
Could we disable database settings for the ws_relay command instead? Are there any settings that affect its behavior that need to be dynamic?

shanemcd replied:
Log level is the main thing I think we'd care about. Maybe we could do that as a short-term bandaid, but it really suck if we declared that our settings just don't work with async code.

AlanCoding commented on awx/main/consumers.py:
```diff
@@ -176,9 +191,20 @@ async def receive_json(self, data):
+
+                await self.channel_layer.group_send(
+                    settings.BROADCAST_WEBSOCKET_GROUP_NAME,
+                    {"type": "consumer.subscribe", "groups": list(new_groups), "origin_channel": self.channel_name},
```
In your architecture diagram, in the new system, in step 2 you mention that a "channel" is created and added to the groups the client subscribed to. This is a distinct different from the former implementation. Is this what's doing that? I see another subscribe action in the management command. I'm just trying to zone in on what the real new things are here.

shanemcd replied:
This is what _sends_ the subscription requests to the task systems. 

That subscription request is received by wsrelay (running on the task system), which creates a new channel:

https://github.com/ansible/awx/blob/3b179cfbc287d6e8041a22b4381b9b485c044d46/awx/main/wsrelay.py#L171-L172

Then we receive messages for the group the originating client is interested in and _relay_ them back to the websocket connection the subscription request came from:

https://github.com/ansible/awx/blob/3b179cfbc287d6e8041a22b4381b9b485c044d46/awx/main/wsrelay.py#L174-L183

shanemcd committed:
WIP

relrod commented on awx/main/consumers.py:
```diff
@@ -176,9 +191,20 @@ async def receive_json(self, data):
             for group_name in new_groups_exclusive:
                 await self.channel_layer.group_add(group_name, self.channel_name)
+
+                await self.channel_layer.group_send(
```
Just making a note for my future self: I think this line is meant to be de-dented one level.

relrod committed:
Fix BROADCAST_WEBSOCKET_PORT for Kube dev

- `settings/minikube.py` gets imported conditionally, when the
  environment variable `AWX_KUBE_DEVEL` is set. In this imported file,
  we set `BROADCAST_WEBSOCKET_PORT = 8013`, but 8013 is only used in the
  docker-compose dev environment. In Kubernetes environments, 8052 is
  used for everything. This is hardcoded awx-operator's ConfigMap.

- Also rename `minikube.py` because it is used for every kind of
  development Kube environment, including Kind.

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
wsrelay is moved to task now, remove from web supervisor

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
add makefile target to load dev image into Kind

Signed-off-by: Rick Elrod <rick@elrod.me>

fosterseth commented on awx/main/wsrelay.py:
```diff
@@ -148,39 +142,79 @@ async def run_loop(self, websocket: aiohttp.ClientWebSocketResponse):
+                    msg = await asyncio.wait_for(self.channel_layer.receive(consumer_channel), timeout=10)
+                except asyncio.TimeoutError:
+                    current_subscriptions = self.producers[name]["subscriptions"]
+                    if len(current_subscriptions) == 0:
```
not sure we are guaranteed to get a proper unsubscribe message from web container (it relies on a graceful websocket disconnect)

I think we'll want to code in a limit here (60 minutes worth of consecutive timeouts?), otherwise this async task may live forever

relrod replied:
> it relies on a graceful websocket disconnect

I followed some upstream code paths, and I haven't been able to convince myself that this is true.

Twisted will call its disconnect code and the callback takes a `wasClean` parameter to determine if the disconnect was clean or not. Daphne uses 'autobahn' for websockets, and particular autobahn's Twisted integration. Ultimate autobahn ends up calling the protocol's `onClose()` after a few layers of indirection.

https://github.com/django/daphne/blob/d59c2bd4242f81206f5aaa04832d7aba898b4cd6/daphne/ws_protocol.py#L158-L177

and Daphne doesn't seem to discriminate on `wasClean`. It pushes a `websocket.disconnect` which will resolve to this: https://github.com/django/channels/blob/ec8ccda38ae6e3a39e879a62a8ece9dd8d30c772/channels/generic/websocket.py#L226
which ends up calling the consumer's `disconnect()`.

I might have missed something, but from reading the code, I _think_ we're okay here.

relrod committed:
dedent a block that was clearly meant to be de-dented

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Merge branch 'devel' into wsrelay

relrod committed:
Remove some debug code and modify logging a bit

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
WIP: Make wsrelay listen for pg_notify heartbeat

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Merge branch 'devel' into wsrelay

relrod committed:
Start of heartbeet daemon

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
fix merge from devel - wsbroadcast -> wsrelay

Signed-off-by: Rick Elrod <rick@elrod.me>

AlanCoding commented on docs/websockets.md:
```diff
@@ -65,14 +65,14 @@ This section will specifically discuss deployment in the context of websockets a
 | `uwsgi`      | listens on port 8050, handles API requests |
 | `daphne`      | listens on port 8051, handles websocket requests |
-| `wsbroadcast`   | no listening port, forwards all group messages to all cluster nodes |
+| `wsrelay`   | no listening port, forwards all group messages to all cluster nodes |
```
> forwards all group messages to all cluster nodes

Request clarification here - would it be more accurate to say that it forwards all group messages to "all web nodes"

Perhaps it would be good to add something to the doc about how something qualifies as a web node, considering that it is different from the `Instance` registration mechanism.

I'm not even sure about the "all group messages" part. I thought I remember something about the web nodes declaring what groups they have clients subscribed to, so only messages in certain groups/channels need to be sent to certain web nodes.

But "no listening port" still sounds right!

relrod replied:
I'm going to resolve this comment to get it off my screen. I will be rewriting (or heavily modifying) `docs/websockets.md` before this PR is ever ready for review. :)

AlanCoding commented on requirements/requirements.in:
```diff
@@ -35,6 +35,7 @@ openshift
 pexpect==4.7.0 # see library notes
 prometheus_client
 psycopg2
+psycopg
```
if _possible_ I would like a comment to state what uses this

relrod replied:
Can do. For the time being I've been kind of using this branch as my playground for working on this stuff, so I'd say don't review things in this branch super closely right now.

relrod replied:
Done in 9755825.

relrod committed:
Prevent looping issue when task/web share a Redis

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Add comment for new psycopg dep

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
add wsrelay to tower-processes

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Remove auto-reconnect logic from wsrelay

We no longer need to do this from wsrelay, as it will automatically try
to reconnect when it hears the next beacon from heartbeet.

This also cleans up the logic for what we do when we want to delete a
node we previously knew about.

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
[wsrelay] Copy the message payload before we relay

We internally manipulate the message payload a bit (to know whether we
are originating it on the task side or the web system is originating
it). But when we get the message, we actually get a reference to the
dict containing the payload.

Other producers in wsrelay might still be acting on the message and
deciding whether or not to relay it. So we need to manipulate and send a
*copy* of the message, and leave the original alone.

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
[wsrelay] attempt to standardize logging levels

This needs some work, but it's a start.

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Merge branch 'devel' into wsrelay

fosterseth committed:
Send subsystem metrics via wsrelay (#13333)

Works by adding a dedicated producer in wsrelay that looks for
local django channels message with group "metrics". The producer
sends this to the consumer running in the web container.

The consumer running in the web container handles the message by
pushing it into the local redis instance.

The django view that handles a request at the /api/v2/metrics
endpoint will load this data from redis, format it, and return the
response.

shanemcd commented on awx/main/wsrelay.py:
```diff
@@ -0,0 +1,307 @@
+        # the "metrics" group is not subscribed to in the typical fashion, so we
+        # just explicitly create it
+        producer = self.event_loop.create_task(self.run_producer("metrics", websocket, "metrics"))
+        self.producers["metrics"] = {"task": producer, "subscriptions": {"metrics"}}
```
@fosterseth I might be missing something here but is this kind of looks like we are only running one metrics producer and then overwriting it every time we establish a new connection. Look below where we create the "normal" producer, we scope it to the name of the remote connection:

```
name = f"{self.remote_host}-{group}"
```

I think what you want might want this to be something like:
```suggestion
        self.producers[f"{self.remote_host}-metrics"] = {"task": producer, "subscriptions": {"metrics"}}
```

relrod committed:
Merge branch 'devel' into wsrelay

relrod committed:
Merge branch 'devel' into wsrelay
